Libplanet/Blockchain/FungibleAssetStateCompleters.cs

```diff
@@ -22,10 +22,10 @@ public static class FungibleAssetStateCompleters<T>
         /// </summary>
         public static readonly FungibleAssetStateCompleter<T> Recalculate =
             (blockChain, blockHash, address, currency) =>
-                blockChain.ComplementBlockStates(blockHash).TryGetValue(
-                    BlockChain<T>.ToFungibleAssetKey(address, currency),
-                    out IValue v
-                ) ? (BigInteger)(Bencodex.Types.Integer)v : 0;
+            {
+                blockChain.ComplementBlockStates(blockHash);
+                return blockChain.GetBalance(address, currency, blockHash);
+            };
```

Comment:
Just curiosity... is this logic the same as before?

Reply:
IMO it's the same.

At first, I thought it will be okay because this passed the tests related to these lines. To add more explanation, before this change, `BlockChain<T>.ComplementBlockStates(HashDigest<SHA256>)` returned `IImmutableDictionary<string, IValue>` (i.e. *string key* → *state*) after complement block states, and `FungibleAssetStateCompleters.Recalculate` tried to get the fungible asset from the value returned. It was because there was a rule in which `BlockChain<T>` uses *BlockStates-StateReferences* model to store states. But it doesn't suit to `IStateStore`, abstraction to get state. For instance, when to apply Merkle Patricia Trie, I'm not sure it's available to present the data structure as `IImmutableDictionary<string, IValue>`.

So I separated it into two steps:

1. Complement block states.
1. Try to get state.

Also, the step to use 0 as default also proceeds in `BlockChain<T>.GetBalance(…)`.
You can see the line [here][blockchain-get-balance-return-line].

[blockchain-get-balance-return-line]: https://github.com/planetarium/libplanet/blob/20f861d66677fd9bc686d040f103c4f4cdeb2ba9/Libplanet/Blockchain/BlockChain.cs#L445

Reply:
You've got my understand. thanks.